aws-common/src/main/java/org/apache/druid/common/aws/LazyFileSessionCredentialsProvider.java

```diff
@@ -21,27 +21,41 @@
 
 import com.amazonaws.auth.AWSCredentials;
 import com.amazonaws.auth.AWSCredentialsProvider;
+import org.checkerframework.checker.nullness.qual.EnsuresNonNull;
+
+import javax.annotation.Nullable;
 
 public class LazyFileSessionCredentialsProvider implements AWSCredentialsProvider
 {
-  private AWSCredentialsConfig config;
-  private FileSessionCredentialsProvider provider;
+  private final AWSCredentialsConfig config;
+
+  /**
+   * The field is declared volatile in order to ensure safe publication of the object
+   * in {@link #getUnderlyingProvider()} without worrying about final modifiers
+   * on the fields of the created object
```

Comment:
I realised that unlikely that many readers of this code will understand this passage without the reference to the thread. Please add `, see https://github.com/apache/incubator-druid/pull/6662#discussion_r237013157` in the end